### PR TITLE
Issue - CRM-21572  Cleared 

### PR DESCRIPTION
**Pull request by jeeskdenny**
Overview
----------------------------------------
_A brief description of the pull request. Try to keep it non-technical._

Before
----------------------------------------
_The current status. Please provide screenshots or gifs ([LICEcap](http://www.cockos.com/licecap/), [SilentCast](https://github.com/colinkeenan/silentcast)) where appropriate._

After
----------------------------------------
_What has been changed. Please provide screenshots or gifs ([LICEcap](http://www.cockos.com/licecap/), [SilentCast](https://github.com/colinkeenan/silentcast)) where appropriate._

Technical Details
----------------------------------------
_If the PR introduces noteworthy technical changes, please describe them here. Provide code snippets if necessary_

Comments
----------------------------------------
_Anything else you would like the reviewer to note_

---

 * [CRM-21572: Cannot store unencoded html in custom fields via API](https://issues.civicrm.org/jira/browse/CRM-21572)

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@twomice this is a PR for a bug you reported.

Note that in order to get this merged it would need a unit test & some discussion as to whether there is any potential security hole here - ie. is the data going in being appropriately checked for injection somewhere.

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by jamienovick** [comment]
@jeeskdenny please can you update your PR with the above information as requested so it can be reviewed fully? Many thanks

@davialexandre @omarabuhussein FYI

File: CRM/Core/BAO/CustomField.php [hunk 1]
```diff
@@ -129,6 +129,7 @@ public static function dataToHtml() {
    * @return CRM_Core_DAO_CustomField
    */
   public static function create(&$params) {
+
```
Comment (xurizaemon) [inline CRM/Core/BAO/CustomField.php]: Code formatting changes should probably go in a separate PR to clarify the actual relevant changes.

I recommend using `git add --patch` to review that changes proposed are relevant before committing.

File: CRM/Core/BAO/CustomValueTable.php [hunk 2]
```diff
@@ -641,6 +641,10 @@ public static function setValues(&$params) {
           $cvParam['file_id'] = $fieldValue['value'];
         }
 
+        if ($cvParam['type'] == 'Memo' && $cvParam['value']) {
+          $cvParam['value'] = html_entity_decode($cvParam['value']);
+        }
+
         if (!array_key_exists($dao->table_name, $cvParams)) {
```
Comment (xurizaemon) [inline CRM/Core/BAO/CustomValueTable.php]: I'm not a fan of special casing field handling based on name, even though this is the existing pattern CiviCRM uses. It complicates what we need to know in order to predict behaviour, and the logic is scattered through multiple locations in CiviCRM.

If as CRM-11532 says CiviCRM's API should now handle input and output without escaping, then the correct fix here seems to be eliminating whatever is encoding that input before storage.

**Comment by eileenmcnaughton** [comment]
I'm going to close this &let you re-open it if you wish to pursue it - there are steps required to get this merged & I suspect you have lost interest. If you still wish to complete this please address the comments & re-open
